### ISSUE TITLE
Size-parametrised are(n)as

### ISSUE THREAD

**Issue by ggreif** [ekmett/thc#1]
More than ten years ago I started working on an idea that would allow shared metadata for a short array of fixed-size objects. (Remnants are here: http://github.com/ggreif/aL4nin).

The idea is to have a c++ (templated) type for GC-ed heap arenas (i.e. fragments of the full GC heap)

``` c++
template <unsigned Anchor, unsigned Compression>
struct Arena
{
  static void* getMeta(void*) { ... }
  ...
};
```

These guys would be allocated with `mmap` such that a sufficient number of least significant bits of the address are zero (e.g. `0x0000007599d30000`). Each arena would be partitioned into a prefix _meta_ and a following _object_ portion. Given a pointer into the latter one can compute a corresponding pointer into the former in a rather cheap manner from the `Anchor` and `Compression` template arguments. (Especially when those are both 2^k.)

Basically one masks away the LSBits corresponding to `Anchor` (obtaining the start of the `Arena`) and divides the offset from the arena to the object by the compression factor to obtain the metadata index. Adding this index to the arena pointer results in the metadata pointer.

In a runtime system many objects can share common meta-data (pvtbl, GC-info etc.) at the cost of some fragmentation in the arenas, because depending on what type of object is to be allocated one needs to keep track of a separate sub-arena which shares the same metadata entries.

**Comment by ekmett** [comment]
It sounds like you managed to reinvent the old LISP-style notion of the BIBOP: [Don't Stop The BIBOP](https://www.cs.indiana.edu/~dyb/pubs/bibop.pdf).

Here I was looking at using a BIBOP style management system to figure out which of the different collectors I'm in, but the downside of doing it on a per object type as you noted, is that you get rather significantly increased fragmentation.

**Comment by ekmett** [comment]
Closing due to this repo changing focus quite severely.
